agent/containermetadata/parse_metadata.go

```
@@ -0,0 +1,151 @@
+)
+
+// parseNetworkMetadata parses the docker.NetworkSettings struct and
+// packages the desired metadata for JSON marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this stage.
+func parseNetworkMetadata(settings *docker.NetworkSettings) (NetworkMetadata, error) {
+	// If network settings are unavailable, users should be notified in case
+	// this was not intended
+	if settings == nil {
+		err := fmt.Errorf("parse network metadata: could not find network settings")
+		return NetworkMetadata{}, err
+	}
+
+	// This metadata is available in two different places in NetworkSettings
+	ipv4Address := settings.IPAddress
+
+	// Network mode is not available for Docker API versions 1.17-1.20
+	networkModeFromContainer := ""
+	if len(settings.Networks) == 1 {
+		for modeFromSettings, containerNetwork := range settings.Networks {
+			networkModeFromContainer = modeFromSettings
+			ipv4Address = containerNetwork.IPAddress
+		}
+	}
+
+	networkMD := NetworkMetadata{
+		networkMode: networkModeFromContainer,
+		ipv4Address: ipv4Address,
+	}
+	return networkMD, nil
+}
+
+// parseDockerContainerMetadata parses the metadata in a docker container
+// and packages this data for JSON marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this stage.
+func parseDockerContainerMetadata(container *docker.Container) DockerContainerMD {
+	if container == nil {
+		seelog.Warn("Failed to parse container metadata: container metadata not available or does not exist")
+		return DockerContainerMD{}
+	}
+
+	// In most cases a container should never lack a config but we check regardless to avoid
+	// nil pointer exceptions (Could occur if there is some error in the docker api call, if the
+	// container we receive has incomplete information)
+	imageNameFromConfig := ""
+	if container.Config != nil {
+		imageNameFromConfig = container.Config.Image
+	} else {
```

This too. if it's an error condition that you want to guard against, you should return an error and not consume it silently. It applies to multiple places in this method. Please fix all of these.